src/mnode/mnode-pastel.cpp

```diff
@@ -603,14 +603,77 @@ bool common_validation(const T& ticket, bool preReg, const std::string& strTnxId
     if (depth > 0)
         return true;
     
-    std::string err;
     if (!pastelTicket->IsValid(false, ++depth)) {
-      throw std::runtime_error(strprintf("The %s ticket with this txid [%s] is invalid - %s", prevTicket, strTnxId, err));
+      throw std::runtime_error(strprintf("The %s ticket with this txid [%s] is invalid", prevTicket, strTnxId));
     }
     
     return true;
 }
 
+void trade_copy_validation(const std::string& nftTnxId, const std::vector<unsigned char>& signature) {
+  if (!masterNodeCtrl.masternodeSync.IsSynced()) {
+    throw std::runtime_error("Can not validate trade ticket as master node is not synced");
+  }
+
+  unsigned short totalCopies{0};
+
+  uint256 txid;
+  txid.SetHex(nftTnxId);
+  auto nftTicket = CPastelTicketProcessor::GetTicket(txid);
+  if (!nftTicket) {
+    throw std::runtime_error(strprintf(
+      "The NFT ticket with txid [%s] referred by this trade ticket is not in the blockchain", nftTnxId));
+  }
+  if (nftTicket->ID() == TicketID::Activate) {
+    auto actTicket = dynamic_cast<const CArtActivateTicket*>(nftTicket.get());
+    if (!actTicket) {
+      throw std::runtime_error(strprintf(
+        "The activation ticket with txid [%s] referred by this trade ticket is invalid", nftTnxId));
+    }
+
+    auto pArtTicket = CPastelTicketProcessor::GetTicket(actTicket->regTicketTnxId, TicketID::Art);
+    if (!pArtTicket) {
+      throw std::runtime_error(strprintf(
+        "The registration ticket with txid [%s] referred by activation ticket is invalid",
+        actTicket->regTicketTnxId));
+    }
+
+    auto artTicket = dynamic_cast<const CArtRegTicket*>(pArtTicket.get());
+    if (!artTicket) {
+      throw std::runtime_error(strprintf(
+        "The registration ticket with txid [%s] referred by activation ticket is invalid",
+        actTicket->regTicketTnxId));
+    }
+
+    totalCopies = artTicket->totalCopies;
+  } else if (nftTicket->ID() == TicketID::Trade) {
+    auto tradeTicket = dynamic_cast<const CArtTradeTicket*>(nftTicket.get());
+    if (!tradeTicket) {
+      throw std::runtime_error(strprintf(
+        "The trade ticket with txid [%s] referred by this trade ticket is invalid", nftTnxId));
+    }
+
+    totalCopies = 1;
+  } else {
+    throw std::runtime_error(strprintf(
+      "Unknown ticket with txid [%s] referred by this trade ticket is invalid", nftTnxId));
+  }
+
+  auto soldCopies{0};
```

Comment:
I think it's better to declare type implicitly. Because it should have same type as totalCopies.

Reply:
ok

Reply:
fixed